### ISSUE TITLE
Cannot set PRAGMA foreign_keys = OFF on migration

### ISSUE THREAD

**Issue by cassioseffrin** [vitusortner/floor#534]
As SQLite has some limitations on ALTER TABLE , we usually have to DROP an existing table and create a new one to achieve a simple migration on a table.

Here is the orientations on the oficial docs: http://www.sqlite.org/faq.html:

The basic Steps:
1```
. CREATE TEMPORARY TABLE t1_backup(a,b);
2. INSERT INTO t1_backup SELECT a,b FROM t1;
3. DROP TABLE t1;
4. CREATE TABLE t1(a,b);
5. INSERT INTO t1 SELECT a,b FROM t1_backup;
6. DROP TABLE t1_backup;

``` 


So in my migration script I have tried to do it as described bellow:
```
  await database.execute("PRAGMA foreign_keys = OFF");
  await database.execute("CREATE TABLE IF NOT EXISTS  temp_table AS SELECT * FROM customer");
  await database.execute("DROP TABLE IF EXISTS customer");
  await database.execute("PRAGMA foreign_keys = ON");
```
  
But in fact the first query in my script is not been executed, consequently  throwing an error when the DROP table is called. (787) SQLITE_CONSTRAINT_FOREIGNKEY. It's because my customer table has some references that i need to ignore when I am dropping it

If I manually change the script in auto generated by build_runner **database.g.dart**   setting the **PRAGMA foreign_keys = OFF** the migration works fine, but i cannot set PRAGMA foreign_keys = ON after migration.

```
 Future<sqflite.Database> open(String path, List<Migration> migrations,
      onConfigure: (database) async {
        await database.execute('PRAGMA foreign_keys = OFF');
      },
```
So the issue is: how can I set PRAGMA foreign_keys OFF AND ON in the migration script?

**Comment by mqus** [comment]
Hm. I think this stems from the issue that all migration steps are run within a transaction (to be able to roll them back if something fails) and changing this setting is silently ignored within transactions (see https://www.sqlite.org/foreignkeys.html#fk_enable).

I'm not sure how to change this. We could provide an optional parameter which disables the foreign keys for migrations and switches back afterwards, regardless of migration success (seems to be useful imho).

PS: Maybe unrelated, but I'm interested: what are the limitations you're running into with `ALTER TABLE`?

**Comment by mqus** [comment]
I was slightly mistaken. The reason for the issue is right, but the transaction is opened within sqflite and not in floor, so we can not fix this in floor.

**Comment by cassioseffrin** [comment]
Thks @mqus .

Look: https://www.sqlite.org/lang_altertable.html
> SQLite supports a limited subset of ALTER TABLE. The ALTER TABLE command in SQLite allows the user to rename a table or to add a new column to an existing table. It is not possible to rename a column, remove a column, or add or remove constraints from a table.

Altering a table structure is very limited in SQLite. The official documentation recommend the steps above in my post.

> I'm not sure how to change this. We could provide an optional parameter which disables the foreign keys for migrations and switches back afterwards, regardless of migration success (seems to be useful imho).
That would be awesome. If you could give me a clue, may I can try to help on it. Once the official docs suggest set PRAGMA foreign_keys = OFF on migrations with FK restrictions It will become a important feature.

**Comment by mqus** [comment]
> It is not possible to rename a column, remove a column, or add or remove constraints from a table.

Where do you read this? The page you linked exactly lines out that you can add, remove and rename columns. Is this somthing that changed with past sqlite versions?

> The only schema altering commands directly supported by SQLite are the "rename table", "rename column", "add column", "drop column" commands shown above

But I get that editing columns involved in constraints (like primary keys, foreign keys, unique columns etc) is not possible.

**Comment by cassioseffrin** [comment]
In the last versions of SQLite they are improving the procedures for altering tables, however we always have users with older versions of android and consequently SQLite ( KitKat and Jelly Bean have SQLite 3.7.4)

We aren't be able to migrate a drop column in older SQLite versions, also we cannot be able to change a FK or index in current SQLite version. The only security change on a table is add an ordinary column. That limitations of SQLite makes the migration very tricky.

I think there are 2 options:
1: Disable the transaction on migration.
2: Create a method to set PRAGMA foreign_keys = OFF 

 Otherwise the migration process will be very imitated.

**Comment by mqus** [comment]
Yes I understand the need. 

Your option for a workaround (2.) would be to do the migrations manually by creating `@Query`s with your actions and manually determining which migrations to run and making sure that it runs before anything else. This is a pain in the ass, yes,  but the only option for now.

 as for 1.: This is only possible if floor does not use the onUpgrade hook of sqflite. I have no clue if this is possible or feasible and how to achieve this, since I haven't looked at it.

**Comment by mqus** [comment]
https://github.com/tekartik/sqflite/issues/624 and https://github.com/tekartik/sqflite/issues/297 have suggestions for solutions useable by floor, all not perfect.

**Comment by cassioseffrin** [comment]
Ok, I have understood your suggestion  workaround (2.).

Tomorrow I'll make a try. I have another idea that consists in create 3 migration scripts instead one: 
1st: foreign_keys = OFF, 
2nd: make the migration;
3rd: set foreign_keys= ON.

May it work without plain @Query's. I will share my tests on this tomorrow. Thks for the great help today @mqus !

**Comment by mqus** [comment]
it will not work with 1, 2, 3, or even thousand migrations scripts because each and every one is executed within the same callback which only is executed within the transaction provided by sqflite (see https://github.com/vitusortner/floor/blob/develop/example/lib/database.g.dart#L76).

I noticed that you could try another workaround like described in the sqflite issues:

Edit the generated sources (if that makes sense in your workflow) and disable the foreignkeys in `onConfigure` and enable them in `onOpen`, as they are executed before and after the `onUpdate` step, respectively and get whole database access.

**Comment by mqus** [comment]
We could improve the workaround by also adding `onConfigure` to the callbacks you can provide to floor.

**Comment by cassioseffrin** [comment]
Yeap, your suggestion to add onConfigure is much better!


Look: https://github.com/tekartik/sqflite/issues/297#issuecomment-548607348
> Maybe you could turn foreign keys off in onConfigure and turn it on later in onOpen (this is called after the transaction is closed)

**Comment by cassioseffrin** [comment]
@mqus  I have tested your commit. However the current floor version  is now depending on flutter above 2.x . Do you have plans to deliver your commit on older floor 0.1X ?

**Comment by mqus** [comment]
I don't think that we will release another 0.x update, we don't even have a branch for that. But I'm not managing the releases here ;)

**Comment by mqus** [comment]
You should be able to set it back to `ON` in the `onOpen` Callback. I added an example in the mentioned PR: 

https://github.com/vitusortner/floor/blob/bc1f945d086fc59f250d996d8c67e5e20ba73e1b/floor/test/integration/database_test.dart#L377-L386

**Comment by mqus** [comment]
the callback order is roughly:
you call database.open() -> onConfigure -> (onCreate OR onUpgrade OR onDowngrade) -> onOpen ->you get your database instance.

in onCreate, the generated schema is applied beforehand by floor.
in onUpgrade, the migrations are applied beforehand by floor.

**Comment by cassioseffrin** [comment]
@mqus  I appreciated so much your help! About 0.1.x version today I have made my tests merging your code into 0.1.x. It's not easy to run it on older versions, a tons of older dependencies needed to match with flutter up2.x.  Better update my project to 2x... it's a painful job unfortunately 

About the next release of floor do you have an idea when that commit will enter in production?

**Comment by cassioseffrin** [comment]
@mqus Do you know if there is a branch of floor related to published 0.19.1 (pub.dev)?

**Comment by cassioseffrin** [comment]
I have done merge of your branch onconfigure-callback on floor 0.19.1. and finally I get this to work including 

I think it could be good to have a branch with this feature in previews version of floor compatible with

**Comment by cassioseffrin** [comment]
I have done merge of your branch onconfigure-callback with floor 0.19.1. and finally I get this to work including floor_generator.

I think it could be nice to have a branch with this feature in previews version of floor compatible with flutter 1.2.x.  If you agree I will make a PR
